source/distributed/tria.cc

```diff
@@ -3775,6 +3776,187 @@ namespace parallel
+        }
+    }
+
+
+
+    template <int dim, int spacedim>
+    void
+    Triangulation<dim, spacedim>::
+    store_quadrant_cell_relations_recursively (const typename dealii::internal::p4est::types<dim>::tree &tree,
+                                               const typename Triangulation<dim,spacedim>::cell_iterator &dealii_cell,
+                                               const typename dealii::internal::p4est::types<dim>::quadrant &p4est_cell)
+    {
+      // find index of p4est_cell in the quadrants array of the corresponding tree
+      int idx = sc_array_bsearch(const_cast<sc_array_t *>(&tree.quadrants),
+                                 &p4est_cell,
+                                 dealii::internal::p4est::functions<dim>::quadrant_compare);
+      if (idx == -1 && (dealii::internal::p4est::functions<dim>::
+                        quadrant_overlaps_tree (const_cast<typename dealii::internal::p4est::types<dim>::tree *>(&tree),
+                                                &p4est_cell)
+                        == false))
+        // this quadrant and none of its children belong to us.
+        return;
+
+      // recurse further if both p4est and dealii still have children,
+      // so that we do not run into a dead end in one of the trees
+      bool p4est_has_children = (idx == -1);
+      if ( p4est_has_children && dealii_cell->has_children() )
+        {
+          //recurse further
+          typename dealii::internal::p4est::types<dim>::quadrant
+          p4est_child[GeometryInfo<dim>::max_children_per_cell];
+
+          for (unsigned int c=0; c<GeometryInfo<dim>::max_children_per_cell; ++c)
+            switch (dim)
+              {
+              case 2:
+                P4EST_QUADRANT_INIT(&p4est_child[c]);
+                break;
+              case 3:
+                P8EST_QUADRANT_INIT(&p4est_child[c]);
+                break;
+              default:
+                Assert (false, ExcNotImplemented());
+              }
+
+          dealii::internal::p4est::functions<dim>::
+          quadrant_childrenv (&p4est_cell, p4est_child);
+
+          for (unsigned int c=0;
+               c<GeometryInfo<dim>::max_children_per_cell; ++c)
+            {
+              store_quadrant_cell_relations_recursively (tree,
+                                                         dealii_cell->child(c),
+                                                         p4est_child[c]);
+            }
+        }
+      else if (!p4est_has_children && !dealii_cell->has_children())
+        {
+          // this active cell didn't change
+          // save tuple into corresponding position
+          store_single_quadrant_cell_relation(tree,
+                                              idx,
+                                              dealii_cell,
+                                              Triangulation<dim,spacedim>::CELL_PERSIST);
+        }
+      else if (p4est_has_children) // && !dealii_cell->has_children()
```

Comment:
remove the comment if it doesn't apply any more

Reply:
or maybe even better, say something like ` // but based on the conditions above, we know that dealii_cell has no children`